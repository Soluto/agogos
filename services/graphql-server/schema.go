package main

import (
	graphql "github.com/graphql-go/graphql"
	"github.com/vektah/gqlparser"
	"github.com/vektah/gqlparser/ast"

	"context"
	"fmt"
	"google.golang.org/grpc"
	"graphql-gateway/generated"
	"graphql-gateway/utils"
	"io"
	"os"
	"time"
)

func getenv(key, fallback string) string {
	value := os.Getenv(key)
	if len(value) == 0 {
		return fallback
	}
	return value
}

var (
	address = getenv("SCHEMA_REGISTRY_URL", "graphql-gateway.registry:81")
)

type schemaResult struct {
	schema *graphql.Schema
	err    error
}

func subscribeToSchema(schemas chan schemaResult) (err error) {
	defer utils.Recovery(&err)

	conn, err := grpc.Dial(address, grpc.WithInsecure())
	if err != nil {
		fmt.Println("error initiating GRPC channel")
		return err
	}
	defer conn.Close()

	gqlSchemaClient := gqlschema.NewGqlSchemaClient(conn)

	stream, err := subscribe(gqlSchemaClient)
	if err != nil {
		fmt.Println("error subscribing to registry", err)
		return err
	}

	for {
		gqlSchemaMessage, err := stream.Recv()

		if err == io.EOF {
			fmt.Println("got EOF")
			break
		}
		if err != nil {
			fmt.Println("error receiving message")
			schemas <- schemaResult{
				schema: nil,
				err:    err,
			}
			return err
		}

		astSchema, err := parseSdl(source{
			name: "schema registry sdl",
			sdl:  gqlSchemaMessage.Schema,
		})
		if err != nil {
			fmt.Println("error parsing SDL")
			schemas <- schemaResult{
				schema: nil,
				err:    err,
			}
			return err
		}

		schema, err := ConvertSchema(astSchema)
		if err != nil {
			fmt.Println("error converting schema")
			schemas <- schemaResult{
				schema: nil,
				err:    err,
			}
			return err
		}

		schemas <- schemaResult{
			schema: schema,
			err:    nil,
		}
	}
	return
}

func subscribe(client gqlschema.GqlSchemaClient) (stream gqlschema.GqlSchema_SubscribeClient, err error) {
	for i := 0; i < 3; i++ {
		stream, err = client.Subscribe(context.Background(), &gqlschema.GqlSchemaSubscribeParams{})
		if err == nil {
			break
		}
		if i+1 < 3 {
			time.Sleep(3000 * time.Millisecond)
		}
	}
	return
}

type source struct {
	name string
	sdl  string
}

func parseSdl(s source) (*ast.Schema, error) {
	schema, err := gqlparser.LoadSchema(&ast.Source{
		Name:  s.name,
		Input: s.sdl,
	})

	if err != nil {
		return nil, err
	}

	return schema, nil
}
