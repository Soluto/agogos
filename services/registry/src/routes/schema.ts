import * as express from "express";
import * as bodyParser from "body-parser";
import sources, { defaultSource } from "../sources-config";
import syncSchema$ from "../sync/sync-service";
import { take } from "rxjs/operators";

const app = express();

const getFromSource = async (source: string, kind: string, res: express.Response) => {
    try {
        const gqlObjects = await sources[source].getGqlObjects(kind);
        res.send(gqlObjects);
    } catch (error) {
        console.warn(`Failed to get ${kind} from source - ${source}`, {
            error
        });
        res.sendStatus(500);
    }
};

const postSource = async (
    source: string,
    kind: string,
    name: string,
    definition: string,
    res: express.Response
) => {
    try {
        await sources[source].registerGqlObject(name, kind, definition);
        res.sendStatus(200);
    } catch (error) {
        console.warn(`Failed to register schema to source - ${source}`, {
            name,
            kind,
            error
        });
        res.sendStatus(500);
    }
};

app
    .use(bodyParser.text())
    .get("/:sourceName/:kind", (req, res) => getFromSource(req.params.sourceName, req.params.kind, res))
    .post("/:sourceName/:kind/:name", (req, res) => {
        const { sourceName, kind, name } = req.params;
        const definition = req.body;
        return postSource(sourceName, kind, name, definition, res);
    })
    .get("/", async (_, res) => res.send(await syncSchema$.pipe(take(1)).toPromise()))
    .post(":kind/:name", (req, res) => {
        const { kind, name } = req.params;
        const definition = req.body;
        return postSource(defaultSource, kind, name, definition, res);
    });

export default app;
