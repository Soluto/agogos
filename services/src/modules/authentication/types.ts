import { ClientOptions } from 'jwks-rsa';

export interface IssuerConfig {
  authority: string;
  jwksConfig?: Partial<ClientOptions>;
  audience?: string | string[];
  authenticatedPaths: string[];
  description?: string;
}

export interface AuthenticationConfig {
  jwt?: Record<string, IssuerConfig>;
  anonymous?: {
    publicPaths: string[];
    rejectAuthorizationHeader?: boolean;
  };
}
