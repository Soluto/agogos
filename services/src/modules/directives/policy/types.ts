import {PolicyArgsObject, PolicyAttachments} from '../../resource-repository/types';

export type Policy = {
    namespace: string;
    name: string;
    args?: PolicyArgsObject;
};

// args here contain the final values after param injection
export type PolicyExecutionContext = {
    namespace: string;
    name: string;
    policyAttachments: PolicyAttachments;
    args?: PolicyArgsObject;
    query?: QueryResults;
};

export type QueryResults = {
    [name: string]: any;
};

export type PolicyExecutionResult = {
    done: boolean;
    allow?: boolean;
    query?: {
        type: string;
        code: string;
    };
};

export type GraphQLArguments = {
    [name: string]: any;
};
