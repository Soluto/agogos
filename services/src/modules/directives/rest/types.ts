export interface KeyValue {
  key: string;
  value: string;
}

export interface RestParams {
  url: string;
  method?: string;
  body?: string;
  bodyArg?: string;
  query?: KeyValue[];
  headers?: KeyValue[];
  timeoutMs?: number;
}
