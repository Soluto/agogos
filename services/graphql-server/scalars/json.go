package scalars

import (
	"strconv"

	"github.com/graphql-go/graphql"
	"github.com/graphql-go/graphql/language/ast"
	"github.com/graphql-go/graphql/language/kinds"
)

func identity(val interface{}) interface{} {
	return val
}

func parseJsonLiteral(literal ast.Value) interface{} {
	switch literal.GetKind() {
	case kinds.StringValue, kinds.BooleanValue:
		return literal.GetValue()

	case kinds.IntValue:
		res, err := strconv.Atoi(literal.GetValue().(string))
		if err != nil {
			return nil
		}
		return res

	case kinds.FloatValue:
		res, err := strconv.ParseFloat(literal.GetValue().(string), 64)
		if err != nil {
			return nil
		}
		return res

	case kinds.ObjectValue:
		if fields, ok := literal.GetValue().([]*ast.ObjectField); ok {
			res := make(map[string]interface{}, len(fields))
			for _, val := range fields {
				res[val.Name.Value] = parseJsonLiteral(val.Value)
			}
			return res
		}

	case kinds.ListValue:
		if list, ok := literal.GetValue().([]ast.Value); ok {
			res := make([]interface{}, len(list))
			for i, val := range list {
				res[i] = parseJsonLiteral(val)
			}
			return res
		}
	}

	return nil
}

var JSON = graphql.NewScalar(graphql.ScalarConfig{
	Name:        "JSON",
	Description: "Standard JSON",

	// Serialize graphql response of JSON type
	Serialize: identity,

	// Parse query variable of JSON type
	ParseValue: identity,

	// Parse query literals of JSON type
	ParseLiteral: parseJsonLiteral,
})
