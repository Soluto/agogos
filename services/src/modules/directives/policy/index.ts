import { PolicyDefinition } from '../../resource-repository';
import { sdl as policySdl, PolicyDirective } from './policy';
import { sdl as policiesSdl, PoliciesDirective, resolvers as policyScalarResolvers } from './policies';
import { sdl as policyQuerySdl, PolicyQueryDirective } from './policy-query';
import PolicyExecutor from './policy-executor';
import { Policy, LoadedPolicy } from './types';

export {
  PolicyExecutor,
  policySdl,
  policiesSdl,
  policyQuerySdl,
  PolicyDirective,
  PoliciesDirective,
  PolicyQueryDirective,
  policyScalarResolvers,
};

export type AuthorizationConfig = {
  policies: PolicyDefinition[];
  policyAttachments?: Record<string, LoadedPolicy>;
  policyExecutor: PolicyExecutor;
  basePolicy?: Policy;
};
