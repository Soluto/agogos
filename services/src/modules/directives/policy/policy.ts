import { GraphQLResolveInfo, GraphQLField, defaultFieldResolver, GraphQLObjectType, FieldNode } from 'graphql';
import { SchemaDirectiveVisitor } from 'graphql-tools';
import { gql } from 'apollo-server-core';
import { RequestContext } from '../../context';
import { GraphQLArguments, Policy } from './types';

export class PolicyDirective extends SchemaDirectiveVisitor {
  visitObject(object: GraphQLObjectType<unknown, RequestContext>) {
    const policy = this.args as Policy;
    const originalResolveObject = object.resolveObject;

    object.resolveObject = async (
      source: unknown,
      fields: Record<string, ReadonlyArray<FieldNode>>,
      context: RequestContext,
      info: GraphQLResolveInfo
    ) => {
      if (!context.ignorePolicies) {
        await context.authorizationConfig.policyExecutor.validatePolicy(policy, source, {}, context, info);
      }

      return originalResolveObject ? await originalResolveObject(source, fields, context, info) : source;
    };
  }

  visitFieldDefinition(field: GraphQLField<unknown, RequestContext>) {
    const policy = this.args as Policy;
    const originalResolve = field.resolve || defaultFieldResolver;

    field.resolve = async (
      source: unknown,
      args: GraphQLArguments,
      context: RequestContext,
      info: GraphQLResolveInfo
    ) => {
      if (!context.ignorePolicies) {
        await context.authorizationConfig.policyExecutor.validatePolicy(policy, source, args, context, info);
      }

      return originalResolve.call(field, source, args, context, info);
    };
  }
}

export const sdl = gql`
  directive @policy(namespace: String!, name: String!, args: JSONObject) on OBJECT | FIELD_DEFINITION
`;
