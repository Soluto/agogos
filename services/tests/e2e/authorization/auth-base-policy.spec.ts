import { GraphQLClient } from 'graphql-request';
import { sleep } from '../../helpers/utility';
import {
  createSchemaMutation,
  createPolicyMutation,
  UpdateSchemasMutationResponse,
  CreatePolicyMutationResponse,
  UpdateBasePolicyMutationResponse,
  createBasePolicyMutation,
} from '../../helpers/registry-request-builder';
import GraphQLErrorSerializer from '../../utils/graphql-error-serializer';
import { Policy } from '../../../src/modules/directives/policy/types';
import * as oidcClients from '../config/oidc/clients-configuration.json';
import { getToken } from '../../helpers/get-token';
import { schema, policies, query } from './auth-base-policy.schema';

let gatewayClient: GraphQLClient;
const registryClient = new GraphQLClient('http://localhost:8090/graphql');

const testCases: [string, { ClientId: string }][] = oidcClients
  .filter(c => c.ClientId.startsWith('e2e-base-policy'))
  .map(c => [c.Description, c]);

describe('Authorization - Base policy', () => {
  const basePolicy: Policy = {
    namespace: 'internal',
    name: 'base_policy',
    args: {
      role: '{jwt?.role}',
    },
  };

  const defaultBasePolicy: Policy = {
    namespace: 'internal',
    name: 'default_base_policy',
  };

  beforeAll(async () => {
    expect.addSnapshotSerializer(GraphQLErrorSerializer);

    const policiesResponse: CreatePolicyMutationResponse = await registryClient.request(createPolicyMutation, {
      policies,
    });
    expect(policiesResponse.updatePolicies.success).toBeTruthy();

    const schemaResponse: UpdateSchemasMutationResponse = await registryClient.request(createSchemaMutation, {
      schema,
    });
    expect(schemaResponse.updateSchemas.success).toBeTruthy();

    const basePolicyResponse: UpdateBasePolicyMutationResponse = await registryClient.request(
      createBasePolicyMutation,
      { basePolicy }
    );
    expect(basePolicyResponse.updateBasePolicy.success).toBeTruthy();

    // Wait for gateway to update before next tests
    await sleep(Number(process.env.WAIT_FOR_REFRESH_ON_GATEWAY) | 1500);
  });

  afterAll(async () => {
    const basePolicyResponse: UpdateBasePolicyMutationResponse = await registryClient.request(
      createBasePolicyMutation,
      { basePolicy: defaultBasePolicy }
    );
    expect(basePolicyResponse.updateBasePolicy.success).toBeTruthy();

    // Wait for gateway to update before next tests
    await sleep(Number(process.env.WAIT_FOR_REFRESH_ON_GATEWAY) | 1500);
  });

  beforeEach(() => {
    gatewayClient = new GraphQLClient('http://localhost:8080/graphql');
  });

  test.each(testCases)('%s', async (_, clientConfig) => {
    const accessToken = await getToken({ clientId: clientConfig.ClientId });
    gatewayClient.setHeader('Authorization', `Bearer ${accessToken}`);
    const result = await gatewayClient.request(query).catch(err => err.response);
    expect(result).toMatchSnapshot();
  });

  test('Anonymous access', async () => {
    const result = await gatewayClient.request(query).catch(err => err.response);
    expect(result).toMatchSnapshot();
  });
});
