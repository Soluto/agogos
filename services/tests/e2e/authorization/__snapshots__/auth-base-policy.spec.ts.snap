// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Authorization - Base policy Anonymous access 1`] = `
{
  "errors": [
    {
      "message": "Unauthorized by policy base_policy in namespace internal",
      "exception": {
        "code": "INTERNAL_SERVER_ERROR"
      }
    },
    {
      "message": "Unauthorized by policy base_policy in namespace internal",
      "exception": {
        "code": "INTERNAL_SERVER_ERROR"
      }
    },
    {
      "message": "Unauthorized by policy override_base_policy in namespace auth_bp",
      "exception": {
        "code": "INTERNAL_SERVER_ERROR"
      }
    }
  ],
  "data": {
    "bp_foo": null,
    "bp_bar": null,
    "bp_baz": null,
    "bp_taz": "TAZ"
  }
}
`;

exports[`Authorization - Base policy JWT pass all policies 1`] = `
Object {
  "bp_bar": "BAR",
  "bp_baz": "BAZ",
  "bp_foo": "FOO",
  "bp_taz": "TAZ",
}
`;

exports[`Authorization - Base policy JWT pass base and regular policy 1`] = `
{
  "errors": [
    {
      "message": "Unauthorized by policy override_base_policy in namespace auth_bp",
      "exception": {
        "code": "INTERNAL_SERVER_ERROR"
      }
    }
  ],
  "data": {
    "bp_foo": "FOO",
    "bp_bar": "BAR",
    "bp_baz": null,
    "bp_taz": "TAZ"
  }
}
`;

exports[`Authorization - Base policy JWT pass none of policies 1`] = `
{
  "errors": [
    {
      "message": "Unauthorized by policy base_policy in namespace internal",
      "exception": {
        "code": "INTERNAL_SERVER_ERROR"
      }
    },
    {
      "message": "Unauthorized by policy base_policy in namespace internal",
      "exception": {
        "code": "INTERNAL_SERVER_ERROR"
      }
    },
    {
      "message": "Unauthorized by policy override_base_policy in namespace auth_bp",
      "exception": {
        "code": "INTERNAL_SERVER_ERROR"
      }
    }
  ],
  "data": {
    "bp_foo": null,
    "bp_bar": null,
    "bp_baz": null,
    "bp_taz": "TAZ"
  }
}
`;

exports[`Authorization - Base policy JWT pass only base policy 1`] = `
{
  "errors": [
    {
      "message": "Unauthorized by policy regular_policy in namespace auth_bp",
      "exception": {
        "code": "INTERNAL_SERVER_ERROR"
      }
    },
    {
      "message": "Unauthorized by policy override_base_policy in namespace auth_bp",
      "exception": {
        "code": "INTERNAL_SERVER_ERROR"
      }
    }
  ],
  "data": {
    "bp_foo": "FOO",
    "bp_bar": null,
    "bp_baz": null,
    "bp_taz": "TAZ"
  }
}
`;

exports[`Authorization - Base policy JWT pass only override policy 1`] = `
{
  "errors": [
    {
      "message": "Unauthorized by policy base_policy in namespace internal",
      "exception": {
        "code": "INTERNAL_SERVER_ERROR"
      }
    },
    {
      "message": "Unauthorized by policy base_policy in namespace internal",
      "exception": {
        "code": "INTERNAL_SERVER_ERROR"
      }
    }
  ],
  "data": {
    "bp_foo": null,
    "bp_bar": null,
    "bp_baz": "BAZ",
    "bp_taz": "TAZ"
  }
}
`;

exports[`Authorization - Base policy JWT pass only regular policy 1`] = `
{
  "errors": [
    {
      "message": "Unauthorized by policy base_policy in namespace internal",
      "exception": {
        "code": "INTERNAL_SERVER_ERROR"
      }
    },
    {
      "message": "Unauthorized by policy base_policy in namespace internal",
      "exception": {
        "code": "INTERNAL_SERVER_ERROR"
      }
    },
    {
      "message": "Unauthorized by policy override_base_policy in namespace auth_bp",
      "exception": {
        "code": "INTERNAL_SERVER_ERROR"
      }
    }
  ],
  "data": {
    "bp_foo": null,
    "bp_bar": null,
    "bp_baz": null,
    "bp_taz": "TAZ"
  }
}
`;

exports[`Authorization - Base policy JWT pass regular and override policy 1`] = `
{
  "errors": [
    {
      "message": "Unauthorized by policy base_policy in namespace internal",
      "exception": {
        "code": "INTERNAL_SERVER_ERROR"
      }
    },
    {
      "message": "Unauthorized by policy base_policy in namespace internal",
      "exception": {
        "code": "INTERNAL_SERVER_ERROR"
      }
    }
  ],
  "data": {
    "bp_foo": null,
    "bp_bar": null,
    "bp_baz": "BAZ",
    "bp_taz": "TAZ"
  }
}
`;
